File: src/main/java/fr/adrienbrault/idea/symfony2plugin/doctrine/intention/DoctrineRepositoryClassConstantIntention.java
package fr.adrienbrault.idea.symfony2plugin.doctrine.intention;

import com.intellij.codeInsight.hint.HintManager;
import com.intellij.codeInsight.intention.PsiElementBaseIntentionAction;
import com.intellij.openapi.editor.Editor;
import com.intellij.openapi.project.Project;
import com.intellij.psi.PsiElement;
import com.intellij.util.IncorrectOperationException;
import com.jetbrains.php.lang.psi.elements.PhpClass;
import com.jetbrains.php.lang.psi.elements.StringLiteralExpression;
import fr.adrienbrault.idea.symfony2plugin.Symfony2ProjectComponent;
import fr.adrienbrault.idea.symfony2plugin.config.SymfonyPhpReferenceContributor;
import fr.adrienbrault.idea.symfony2plugin.doctrine.EntityHelper;
import fr.adrienbrault.idea.symfony2plugin.util.MethodMatcher;
import fr.adrienbrault.idea.symfony2plugin.util.PhpElementsUtil;
import org.apache.commons.lang.StringUtils;
import org.jetbrains.annotations.NotNull;

/**
 * @author Daniel Espendiller <daniel@espendiller.net>
 */
public class DoctrineRepositoryClassConstantIntention extends PsiElementBaseIntentionAction {

    @Override
    public void invoke(@NotNull Project project, Editor editor, @NotNull PsiElement psiElement) throws IncorrectOperationException {
        PsiElement parent = psiElement.getParent();
        if(!(parent instanceof StringLiteralExpression)) {
            return;
        }

        try {
            PhpClass phpClass = EntityHelper.resolveShortcutName(project, ((StringLiteralExpression) parent).getContents());
            if(phpClass == null) {
                throw new Exception("Can not resolve model class");
            }
            PhpElementsUtil.replaceElementWithClassConstant(phpClass, parent);
        } catch (Exception e) {
            HintManager.getInstance().showErrorHint(editor, e.getMessage());
        }
    }

    @Override
    public boolean isAvailable(@NotNull Project project, Editor editor, @NotNull PsiElement psiElement) {
        if(!Symfony2ProjectComponent.isEnabled(psiElement.getProject())) {
            return false;
        }

        PsiElement parent = psiElement.getParent();
        if(!(parent instanceof StringLiteralExpression)) {
            return false;
        }

        String contents = ((StringLiteralExpression) parent).getContents();
        if(StringUtils.isBlank(contents)) {
            return false;
        }

        return null != new MethodMatcher.StringParameterMatcher(parent, 0)
            .withSignature(SymfonyPhpReferenceContributor.REPOSITORY_SIGNATURES)
            .withSignature("Doctrine\\Persistence\\ObjectManager", "find")
            .withSignature("Doctrine\\Common\\Persistence\\ObjectManager", "find") // @TODO: missing somewhere
            .match();
    }

    @NotNull
    @Override
    public String getFamilyName() {
        return "Symfony";
    }

    @NotNull
    @Override
    public String getText() {
        return "Doctrine: use class constant";
    }


}

File: src/main/kotlin/fr/adrienbrault/idea/symfony2plugin/doctrine/intention/DoctrineRepositoryClassConstantIntention.kt
package fr.adrienbrault.idea.symfony2plugin.doctrine.intention

import com.intellij.codeInsight.intention.PsiElementBaseIntentionAction
import com.intellij.openapi.editor.Editor
import com.intellij.openapi.project.Project
import com.intellij.psi.PsiElement
import com.jetbrains.php.lang.psi.elements.StringLiteralExpression
import fr.adrienbrault.idea.symfony2plugin.Symfony2ProjectComponent
import fr.adrienbrault.idea.symfony2plugin.config.SymfonyPhpReferenceContributor
import fr.adrienbrault.idea.symfony2plugin.doctrine.EntityHelper
import fr.adrienbrault.idea.symfony2plugin.util.IdeHelper
import fr.adrienbrault.idea.symfony2plugin.util.MethodMatcher
import fr.adrienbrault.idea.symfony2plugin.util.PhpElementsUtil
import org.apache.commons.lang3.StringUtils

/**
 * @author Daniel Espendiller <daniel@espendiller.net>
 */
open class DoctrineRepositoryClassConstantIntention : PsiElementBaseIntentionAction() {
    override fun invoke(project: Project, editor: Editor?, psiElement: PsiElement) {
        val parent = psiElement.parent as? StringLiteralExpression
            ?: return

        try {
            val phpClass = EntityHelper.resolveShortcutName(project, parent.contents)
                ?: throw Exception("Can not resolve model class")
            PhpElementsUtil.replaceElementWithClassConstant(phpClass, parent)
        } catch (e: Exception) {
            IdeHelper.showErrorHintIfAvailable(editor!!, e.message!!)
        }
    }

    override fun isAvailable(project: Project, editor: Editor?, psiElement: PsiElement): Boolean {
        if (!Symfony2ProjectComponent.isEnabled(project)) {
            return false
        }

        val parent = psiElement.parent as? StringLiteralExpression
            ?: return false

        if (StringUtils.isBlank(parent.contents)) {
            return false
        }

        return MethodMatcher.StringParameterMatcher(parent, 0)
            .withSignature(SymfonyPhpReferenceContributor.REPOSITORY_SIGNATURES)
            .withSignature("Doctrine\\Persistence\\ObjectManager", "find")
            .withSignature("Doctrine\\Common\\Persistence\\ObjectManager", "find") // @TODO: missing somewhere
            .match() != null
    }

    override fun getFamilyName(): String {
        return "Doctrine: use class constant"
    }

    override fun getText(): String {
        return familyName
    }
}
